htdocs/core/lib/functions.lib.php

```diff
@@ -6990,6 +6990,11 @@ function getCommonSubstitutionArray($outputlangs, $onlykey = 0, $exclude = null,
 				$substitutionarray['__RECEPTIONTRACKNUMURL__'] = 'Shipping tracking url';
 			}
 		} else {
+			// can substitute variables for project
+			if (!empty($conf->projet->enabled) && method_exists($object, 'fetch_project') && is_null($object->project)) {
```

Comment:
Because method fetch_project is inside the commonobject, it will exists for nearly all objects, even the one that does not have project on them. This means we will make a fetch_object on object where this does not exists.
I would include also a test if property fk_project exists on object to avoid this.
Note the "subtle" difference because:  "feature was never available or not completely implemented for a given use case" is not similar to a "bug". To avoid risk and side effect, when the limit is border-line, i prefer to this such change into develop.

Reply:
@eldy 
I don't understand your point.
The property "fk_project" is also in commonobject and method "fetch_project" return 0 if "fk_project" is empty.

Reply:
I have think of this need for a long time. And i think we must implement this differently:  
The method getCommonSubstitutionArray is called very frequently with different context and need.
And we can start to introduct subrequest (fetch_projet) each time we need a new substitution data. In future we will need fetch_thirdparty, fetch_contact, xxx. This means we will make a ton of SQL request a lot of time.
So to avoid this i think a different way to implement your need.
You can add a substitution key  
$substitutionarray[__XXX__@lazyload]  = 'path:class:method'

Then inside the method make_substitutions(), we add a loop to scan $substitutionarray:
For each key ending with '@lazyload', we extract the substitution key '__XXX__' and we check inside the $text (the 1st parameter of  make_substitutions), the the string __XXX__ exists.
If no, we replace with ''.
If yes, we can make the 
```
include_once $path;
$tmpobj = new $class($db);
$valuetouseforsubstitution = $tmpobj->$method(__XXX__)
```

And make the replacement of "__XXX__@lazyload" with valuetouseforsubstitution

So the SQL access or even any complex and heavy code to get a subsitution key value is done if and only if needed.

Reply:
@eldy 
If you put "__XXX___@lazyload" instead of "__XXX__", the final user will have to change all templates with the new "__XXX___@lazyload" key in their email templates or documents. How will he know if he needs to add or not "@lazyload" ?

Before calling the fetch_project method, it checks that project hasn't been loaded yet. So the "fetch_project" method (and SQL) is called only one time in all cases.

Furthermore, when you call 'getCommonSubstitutionArray()', you want to make substitution and call immediately after  "make_substitutions" method. So you got the sames SQL requests.